src/libostree/ostree-repo-refs.c

```diff
@@ -792,6 +792,25 @@ _ostree_repo_write_ref (OstreeRepo    *self,
     }
   else
     {
+      g_autoptr(GHashTable) refs = NULL;
+      GHashTableIter hashiter;
+      gpointer hashkey, hashvalue;
+
+      ostree_repo_list_refs (self, ref, &refs, cancellable, error);
+      g_hash_table_iter_init (&hashiter, refs);
+
+      while ((g_hash_table_iter_next (&hashiter, &hashkey, &hashvalue)))
+        {
+          if (strcmp (ref, (char *)hashkey) != 0)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Specified ref %s already exists as a folder", ref);
+              goto out;
+            }
+        }
+
+      glnx_shutil_rm_rf_at (dfd, ref, cancellable, error);
+
       if (!write_checksum_file_at (self, dfd, ref, rev, cancellable, error))
```

Comment:
So after playing around with the code a bit, the problem is say, if I had a many levels of empty subfolders but wanted to create a new ref, unlinkat would catch the fact that there are existing subfolders (although empty) and return ENOTEMPTY. So if I wanted to delete all of them, I would have to look in each subfolder (recursively) and proceed to delete them one by one. That's why initially I iterated through the existing refs to check for the name instead and rm_rf all the folders if none of them are in use, and it would not require recursive check and deletion. Is it ok if I go back to that? Or would you rather I do recursive check and deletion?